### DISCUSSION TITLE
Compatible WiFi+BT cards for OPi5 & OPi5 Plus

### DISCUSSION THREAD

**Discussion by nobara256** [Joshua-Riek/ubuntu-rockchip#192] [General]
Device: OPi5 & OPi5 Plus
System: v1.17

OPi5 seems to support only a few WiFi cards including the RADXA WiFi modules with Broadcom chips and some Intel WiFi modules.

1. Broadcom modules have SDIO+UART interface and their bandwidth is quite low.
2. Intel modules have PCIE+USB interface so they should have better bandwidth and compatibility.

I tried my laptop's AX201 card and it wasn't even detected in `lspci`. It seem some Intel WiFi cards use Intel's CNVio interface which can't be used with other chip-sets, AX201 is one of them. I ordered a Intel 7265NGW and it works properly out of the box.

Intel AX200 & AX210 seems to work on paper but I doubt their bluetooth functionality would work properly on kernel 5.10.

**Comment by Joshua-Riek** [comment]
I will look at some of the WiFi + BT cards from the rkwifibt repo available from Rockchip. I'll try to add the Realtek firmware, and it should be easy to get wifi working for these cards. Bluetooth can be a bit more challenging.

repo: https://gitlab.com/rk3588_linux/linux/external/rkwifibt/-/tree/linux-5.10-gen-rkr3.5/

**Comment by rich-nahra** [comment]
I have an AX210 E-KEY based WiFI6/BT board.  WiFi works out of the box.  I was able to get Bluetooth working by copying the `intel` directory `/usr/lib/firmware/intel ` off OrangePI Debian Bookworm SDCard. 

`sudo cp -r /media/{username}/opi_root/usr/lib/firmware/intel /usr/lib/firmware/`

**Comment by Xeue** [comment]
Just reviving this thread, @Joshua-Riek I did have the intel AX210 working out of the box in one of your older builds, however in 31 and 32 it isn't working at all. It looks like it's trying to load cfg80211 and failing? I tried the backport driver, which made no differance.
I'm seeing logs like this in dmesg:
[    5.344940] iwlwifi: Unknown symbol ieee80211_channel_to_frequency (err -2)
I've seen in a few places about the cfg80211 module being loaded incorectly as part of the system being an issue rather than as a module.
If I run which on cfg80211 I see lots of links to locations in the linux headers folders and nothing else.

**Reply by squassina** [reply]
This is unfortunate.. did you do a fresh install or update?

I've been updating my system for a while and it's working fine.

But I think I stopped at build 28, as per uname -a:

`Linux desktop 5.10.160-rockchip #28 SMP Thu Dec 28 14:57:13 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux`

**Reply by squassina** [reply]
Just installed the latest version. I had to delete the pnvm file for Wi-Fi to work:
`$ sudo rm /lib/firmware/iwlwifi-ty-a0-gf-a0.pnvm`

Can you git it a shot and report back?

**Reply by Xeue** [reply]
From a fresh install I have the same issue, I need to remove the pnvm file for wifi to work at all. Then when removed it will connect, but won't roam at all. after disconnecting I have to turn the wifi on and off for it to come back...
